docs/housing-counselor-tool.md

```diff
@@ -0,0 +1,228 @@
+        "agc_STATUS": "A",
+        "city": "Springfield",
+        "languages": ["English", "Spanish"],
+        "faithbased": "N",
+        "mailingstatecd": "MI",
+        "email": "counselor@example.org",
+        "fax": "999-888-7777",
+        "phone1": "111-222-3333",
+        "distance": 5.430720023569205,
+        "phone2": " ",
+        "agc_ADDR_LATITUDE": 42.442658,
+        "agcid": "80790",
+        "agc_SRC_CD": "HUD",
+        "nme": "Example Community Housing Services",
+        "migrantwkrs_IND": "N",
+        "parentid": "82772",
+        "services": ["Mortgage Delinquency and Default Resolution Counse", "Home Improvement and Rehabilitation Counseling", "Pre-purchase Counseling", "Pre-purchase Homebuyer Education Workshops", "Rental Housing Counseling"],
+        "counslg_METHOD": "Face to Face Counseling,Group Counseling,Phone Counseling",
+        "county_NME": "",
+        "mailingadr1": "1234 N. Example St.",
+        "statecd": "MI",
+        "mailingadr2": " ",
+        "mailingzipcd": "48219-8888",
+        "state_NME": "Michigan",
+        "agc_ADDR_LONGITUDE": -83.28329,
+        "colonias_IND": "Y"
+    }]
+}
+```
+
+Save the resulting JSON files on the Jenkins job workspace, in a `jsons` directory, e.g. `jsons/12345.json`.
+
+
+### Generate HTML files
+
+When enabled, the `MAKE_HTML` step generates an HTML page of housing counselor results, including styles,
+for each file created in the previous step.
+It saves them in an `htmls` directory on the Jenkins job workspace.
+This step is enabled by default.
+
+This step calls the [`hud_generate_html`](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/housing_counselor/management/commands/hud_generate_html.py) management command,
+which calls HTML generation code in [`generator.py`](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/housing_counselor/generator.py).
+Django renders the [`housing_counselor/pdf_selfcontained.html`](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/housing_counselor/templates/housing_counselor/pdf_selfcontained.html) template with the housing counselor data from each JSON file.
+We save the resulting HTML files on the Jenkins job workspace, in a `htmls` directory, e.g. `htmls/12345.html`.
+
+
+### Generate PDFs
+
+When enabled, the `MAKE_PDF` step generates a PDF of each file created in the previous step.
+It saves them in a `pdfs` directory on the Jenkins job workspace.
+This step is enabled by default.
+
+This step uses a HTML to PDF conversion command line tool.
+We run the conversion on each file in the `htmls` directory, generating a PDF version of each, e.g. `pdfs/12345.pdf`.
+
+
+### Upload to S3
+
+When enabled, the `UPLOAD_TO_S3` step uploads the contents of the `jsons` and `pdfs` directories to an Amazon S3 bucket where it can be accessed by consumerfinance.gov.
+This step is enabled by default.
+
+The files are publicly accessible, e.g.:
+
+ - http://files.consumerfinance.gov/a/assets/hud/jsons/12345.json
```

Comment:
We have a redirect set up on http:// requests to files.cf.gov to https://, and files.cf.gov now supports HTTPS, so these two items can be combined into a single one: https://files.consumerfinance.gov/a/assets/hud/jsons/12345.json

Reply:
Interesting. Does that mean this code is out of date? https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/s3utils.py#L4

Reply:
Nice find! I'd forgotten about that code.

If I recall, I think the idea behind that code was to support generating URLs for arbitrary files on S3, where the S3 bucket name is defined by the `AWS_STORAGE_BUCKET_NAME` Django setting. In theory one could run cfgov-refresh with that setting set to something other than "files.consumerfinance.gov" (for example, if/when build gets set up with its own S3 bucket).

Currently, however, the only use of that code is in the housing counselor logic [here](https://github.com/cfpb/cfgov-refresh/blob/50b9a4458f18836492b6147fb3cfbc9f83645537/cfgov/housing_counselor/views.py#L48) -- but this seems wrong, because here we don't want the URLs to be dependent on that setting, because the files only ever exist in one place (on files.cf.gov). Even if other instances of cf.gov (like build) that had their own S3 buckets for, say, Wagtail uploads, we'd still want their housing counselor page to be pulling files from files.cf.gov.

Given that, maybe it makes sense to change [`settings.HOUSING_COUNSELOR_S3_PATH_TEMPLATE`](https://github.com/cfpb/cfgov-refresh/blob/fb16e906bc4935669f880c270ccf4e32b930b068/cfgov/cfgov/settings/base.py#L324) to be a full path like `https://files.consumerfinance.gov/a/assets/hud/{format}s/{zipcode}.{format}`, thus ensuring that every cf.gov instance everywhere would always use the right bucket. This would let us remove this s3utils file entirely, as it's not used anywhere else.

I think we'll need to do this anyway if/when build gets its own bucket, because otherwise the housing counselor download links would all be wrong... @schbetsy what do you think? I can open a PR with this change if you think it makes sense.

Reply:
PR for this change: https://github.com/cfpb/cfgov-refresh/pull/4759